### PR TITLE
Use lib conditional types for type facts if possible

### LINKED ISSUE #22137: Type narrowing doesn't work for class using generic Readonly<Props>
**TypeScript Version:**  "2.8.0-dev.20180221"
Not in "2.7.0-rc"

**Search Terms:** type narrowing class generics function method call props readonly

**Code**

```ts
// Loose function or method. Same result.
declare function bar(thing: string): void;

interface Base {
    thing?: string;
}
class Example<Props extends Base> {
    props: Readonly<Props>;
    mutableProps: Props;
    base: Readonly<Base>;

    foo() {
        if (this.props.thing) {
            this.props.thing.toString(); // No error
            bar(this.props.thing); // Error!
        }
        if (this.mutableProps.thing) {
            bar(this.mutableProps.thing); // No error
        }
        if (this.base.thing) {
            bar(this.base.thing); // No error
        }
    }
}
```

**Expected behavior:**

No error.

**Actual behavior:**

TypeScript error, but only with that combination of generics, `Readonly`, a class, and calling a function. Remove any one of those factors and it works as expected. Some examples of that are also shown above.

```
Argument of type 'P["thing"]' is not assignable to parameter of type 'string'.
  Type 'string | undefined' is not assignable to type 'string'.
    Type 'undefined' is not assignable to type 'string'.
```

**Playground Link:**

Note: Turn on StrictNullChecks.

https://www.typescriptlang.org/play/#src=%0D%0A%2F%2F%20Loose%20function%20or%20method.%20Same%20result.%0D%0Adeclare%20function%20bar(thing%3A%20string)%3A%20void%3B%0D%0A%0D%0Ainterface%20Base%20%7B%0D%0A%20%20%20%20thing%3F%3A%20string%3B%0D%0A%7D%0D%0Aclass%20Example%3CProps%20extends%20Base%3E%20%7B%0D%0A%20%20%20%20props%3A%20Readonly%3CProps%3E%3B%0D%0A%20%20%20%20mutableProps%3A%20Props%3B%0D%0A%20%20%20%20base%3A%20Readonly%3CBase%3E%3B%0D%0A%0D%0A%20%20%20%20foo()%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(this.props.thing)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20this.props.thing.toString()%3B%20%2F%2F%20No%20error%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20bar(this.props.thing)%3B%20%2F%2F%20Error!%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20if%20(this.mutableProps.thing)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20bar(this.mutableProps.thing)%3B%20%2F%2F%20No%20error%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20if%20(this.base.thing)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20bar(this.base.thing)%3B%20%2F%2F%20No%20error%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A

**Related Issues:**

None that I am aware of. This showed up when updating TypeScript, so it is probably a recent regression.

### PR DESCRIPTION
**Pull request by weswigham**
This causes generics to be narrowed as one may expect, similarly to nongeneric types.

I've not accepted the updated baselines (so this PR should fail), as there are some changes I do not like caused by #22347.

Fixes #22137.

These types are definitely more precise than the fact system - in fact, they even found a bug in fourslash (without strict null checks on!).

### CODE REVIEW

**Comment by weswigham** [comment]
Constraints and intersections as mentioned in the issue in the OP fix most problems (if ugly), but the addition of real, visible constraints causes us to infer literal types in some places just because CFA has constrained a type to some domain, which _also_ is probably undesirable. Traded one set of issues for another, really.

**Comment by zdila** [comment]
Just for the reference: https://stackoverflow.com/questions/50819907/typescript-tagged-union-in-generic

More complex usecase: https://bit.ly/2JLbSkY

**Comment by alvis** [comment]
@weswigham any update?

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -2705,7 +2705,7 @@ namespace ts {
     }
 
     function asToken<TKind extends SyntaxKind>(value: TKind | Token<TKind>): Token<TKind> {
-        return typeof value === "number" ? createToken(value) : value;
+        return typeof value === "number" ? createToken(value) : value as any; // TODO: FIXME
```
Comment (RyanCavanaugh) [inline src/compiler/factory.ts]: What's this about?

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -3299,7 +3299,7 @@ Actual: ${stringify(fullActual)}`);
                 return ts.createTextSpanFromRange(ranges[index]);
             }
             else {
-                this.raiseError("Supplied span index: " + index + " does not exist in range list of size: " + (ranges ? 0 : ranges.length));
+                this.raiseError("Supplied span index: " + index + " does not exist in range list of size: " + (!ranges ? 0 : ranges.length));
```
Comment (RyanCavanaugh) [inline src/harness/fourslash.ts]: 🤣

**Comment by RyanCavanaugh** [comment]
Would like to review this at the design meeting first. cc @DanielRosenwasser

**Comment by weswigham** [comment]
This has inspired many a conversation over the years, however as-is we know we have no shot of merging this. Conditional types have too many drawbacks for it to be OK to introduce them in control flow (like being very hard to relate correctly). We may reconsider, however, in a post-negated-types world, assuming the performance can be tuned, using aliases based on negated types instead. In any case, this PR is old, crusty, and never going to be updated (it'd be far better to make a new PR in such a post-negated-types world), so it's time to close it.

**Pull request closed without merging**